File: Coding_0801 Heap_And_Priority_Queue_And_Intervals/이강훈/류제영/2번.java
# https://www.acmicpc.net/problem/1374

'''
1. 아이디어 :
    우선순위 큐에 끝나는 시간을 넣어, 시작시간보다 작거나 같으면 큐를 빼줌
2. 시간복잡도 :
    O(n)
3. 자료구조 :
    우선순위 큐
'''

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        List<String[]> list = new ArrayList<>();

        int N = Integer.parseInt(br.readLine());

        while (N-- > 0) {
            list.add(br.readLine().split(" "));
        }

        list.sort((o1, o2) -> {
            if (o1[1].equals(o2[1])) return Integer.parseInt(o1[2]) - Integer.parseInt(o2[2]);

            return Integer.parseInt(o1[1]) - Integer.parseInt(o2[1]);
        });

        int answer = 0;
        Queue<Integer> queue = new PriorityQueue<>();
        for (String[] strings : list) {
            while (!queue.isEmpty() && Integer.parseInt(strings[1]) >= queue.peek()) {
                queue.poll();
            }

            queue.offer(Integer.parseInt(strings[2]));
            answer = Math.max(answer, queue.size());
        }

        System.out.println(answer);
    }
}

File: Coding_0929 Two_Pointers_And_Sliding_Window/이강훈/최원준/2번.py
# https://www.acmicpc.net/problem/

'''
1. 아이디어 :
    dp 배열 두개를 이용한다.
2. 시간복잡도 :
    O(n) * 3
3. 자료구조 :
    배열
'''

n = int(input())
cmin = list(map(int, input().split()))
cmax = cmin.copy()
for i in range(n - 1):
    x, y, z = list(map(int, input().split()))
    cmin = [min(cmin[0], cmin[1]) + x, min(cmin) + y, min(cmin[1], cmin[2]) + z]
    cmax = [max(cmax[0], cmax[1]) + x, max(cmax) + y, max(cmax[1], cmax[2]) + z]
print(max(cmax), min(cmin))

